### PR TITLE
docs: move collections APIs to ibis/expr/types and document remaining APIs

### LINKED ISSUE #1632: docs: add API docs for complex data types: arrays and maps
cc @missing-semicolon

### PR DESCRIPTION
**Pull request by cpcloud**
This PR moves collections APIs from ibis/expr/api.py into ibis/expr/types and documents all of the remaining collections APIs across arrays, structs and maps. Closes #1632.

### CODE REVIEW

File: ibis/expr/types/arrays.py [hunk 1]
```diff
@@ -9,23 +9,213 @@
...
+
+        Examples
+        --------
+        >>> import ibis
+        >>> a = ibis.array([1, 2, 3])
+        >>> a.length()
+        int64 = ArrayLength
+          value: array<int8> = [1, 2, 3]
+        """
+        import ibis.expr.operations as ops
```
Comment (kszucs) [inline ibis/expr/types/arrays.py:35]: Can we import it on a module level?

**Comment by kszucs** [comment]
I'm really in favor of this change since it reduces the number of indirections we have in the codebase making easier to both maintenance and contribution.

File: ibis/expr/types/arrays.py (hunk 1 above)
Comment (cpcloud) [reply ibis/expr/types/arrays.py:35]: No, because of circular imports.

**Comment by jreback** [comment]
this looks backwards compatible, right?

**Comment by cpcloud** [comment]
> this looks backwards compatible, right?

Yep. No tests were changed and they're all still passing. I'm not sure if there's anything more we can do to ensure compatibility.

**Comment by jreback** [comment]
> > this looks backwards compatible, right?
> 
> Yep. No tests were changed and they're all still passing. I'm not sure if there's anything more we can do to ensure compatibility.

nope lgtm

**Review by jreback — APPROVED** [review]

**Pull request merged**